cpp/src/IceBox/Admin.cpp

```diff
@@ -7,44 +7,75 @@
 //
 // **********************************************************************
 
-#include <Ice/Application.h>
+#include <Ice/Ice.h>
 #include <Ice/SliceChecksums.h>
 #include <Ice/ConsoleUtil.h>
 #include <IceUtil/Options.h>
 #include <IceBox/IceBox.h>
 
 using namespace std;
-using namespace Ice;
 using namespace IceInternal;
 
-class Client : public Ice::Application
-{
-public:
+int run(Ice::StringSeq&);
 
-    void usage();
-    virtual int run(int, char*[]);
-};
+//
+// Global variable for destroyCommunicator
+//
+Ice::CommunicatorPtr communicator;
 
-#ifdef _WIN32
+//
+// Callback for CtrlCHandler
+//
+void
+destroyCommunicator(int)
+{
+    communicator->destroy();
+}
 
 int
+#ifdef _WIN32
 wmain(int argc, wchar_t* argv[])
-
 #else
-
-int
 main(int argc, char* argv[])
-
 #endif
 {
-    Client app;
-    return app.main(argc, argv);
+    int status = 0;
+
+    try
+    {
+        //
+        // CtrlCHandler must be created before the communicator or any other threads are started
+        //
+        Ice::CtrlCHandler ctrlCHandler;
+
+        //
+        // CommunicatorHolder's ctor initializes an Ice communicator,
+        // and it's dtor destroys this communicator.
+        //
+        Ice::CommunicatorHolder ich(argc, argv);
+        communicator = ich.communicator();
+
+        //
+        // Destroy communicator on Ctrl-C
+        //
+        ctrlCHandler.setCallback(&destroyCommunicator);
+
+        Ice::StringSeq args = Ice::argsToStringSeq(argc, argv);
+        status = run(args);
+    }
+    catch(const std::exception& ex)
+    {
+        cerr << ex.what() << endl;
```

Comment:
You must use consoleErr, this ensure message gets converted to the console encoding